src/plugins/intel_npu/src/backend/src/zero_infer_request.cpp

```diff
@@ -16,131 +16,112 @@ using namespace intel_npu;
-    const std::unordered_map<std::string, ZeroExecutor::ArgumentDescriptor>& executorInputDescriptors,
-    const std::unordered_map<std::string, ZeroExecutor::ArgumentDescriptor>& executorOutputDescriptors) {
-    std::set<size_t> batch_size;
+    auto checkDescriptorsUseCandidateBatchSize = [candidateBatchSize](const std::vector<IODescriptor>& descriptors) {
+        for (const IODescriptor& descriptor : descriptors) {
+            OPENVINO_ASSERT(descriptor.shapeFromIRModel.has_value(),
+                            "Missing value for the \"shapeFromIRModel\" attribute, I/O descriptor");
 
-    Logger logger("getBatchSize", Logger::global().level());
+            const ov::PartialShape& shapeFromCompiler = descriptor.shapeFromCompiler;
+            const ov::PartialShape& shapeFromIRModel = *descriptor.shapeFromIRModel;
 
-    for (const std::string& inputName : metadata.inputNames) {
-        auto batchSizeForNode =
-            getBatchSizeForNode(metadata.parameters.at(inputName), executorInputDescriptors.at(inputName));
-
-        if (batchSizeForNode.has_value()) {
-            batch_size.insert(*batchSizeForNode);
-        } else {
-            return std::nullopt;
-        }
-    }
+            if (shapeFromCompiler.is_dynamic() || shapeFromCompiler.rank().get_length() == 0 ||
+                *shapeFromCompiler.begin() != DEFAULT_BATCH_SIZE) {
+                return false;
+            }
 
-    for (const std::string& outputName : metadata.outputNames) {
-        if (!executorOutputDescriptors.count(outputName)) {
-            OPENVINO_THROW("Invalid graph output descriptor key: " + outputName);
+            if (!descriptor.isStateInput && !descriptor.isStateOutput && !descriptor.isShapeTensor) {
+                if (shapeFromIRModel.is_dynamic() || shapeFromIRModel.rank().get_length() == 0 ||
+                    *shapeFromIRModel.begin() != candidateBatchSize) {
+                    return false;
+                }
+            }
         }
-        auto batchSizeForNode =
-            getBatchSizeForNode(metadata.results.at(outputName), executorOutputDescriptors.at(outputName));
 
-        if (batchSizeForNode.has_value()) {
-            batch_size.insert(*batchSizeForNode);
-        } else {
-            return std::nullopt;
-        }
-    }
+        return true;
+    };
 
-    if (batch_size.size() != 1) {
-        logger.info("Batching works only when we have the same batch size for all tensors!");
-        return std::nullopt;
+    if (!checkDescriptorsUseCandidateBatchSize(metadata.inputs) ||
+        !checkDescriptorsUseCandidateBatchSize(metadata.outputs)) {
+        _logger.info("Batching on the plugin is not used, batching is handled by the compiler");
+        return DEFAULT_BATCH_SIZE;
     }
 
-    auto it = batch_size.begin();
-    if (*it) {
-        return *it;
-    }
+    _logger.info("Batching is handled by the plugin");
 
-    return std::nullopt;
+    return candidateBatchSize;
 }
```

Comment:
I admit refactoring this one wasn't necessary. I ended up writting this in another context, then I've decided to move it here and replace the older function. I'm leaning towards this implementation for several minor reasons:
* Previously, the logger verbosity level was not handled correctly.
* The current implementation is not relying on Level Zero specific objects.
* Fail fast approach - slightly more efficient I believe.